app/src/main/java/org/jellyfin/mobile/AppPreferences.kt

```diff
@@ -52,6 +54,28 @@ class AppPreferences(context: Context) {
             }
         }
 
+    var downloadLocation: String
+        get() {
+            val defaultStorage = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).absolutePath
+            val savedStorage = sharedPreferences.getString(Constants.PREF_DOWNLOAD_LOCATION, null) ?: defaultStorage
+            return with(File(savedStorage)) {
+                parentFile?.run {
+                    if (exists() && isDirectory) return@with savedStorage
+                }
```

Comment:
What does this check? Should it check the file itself?

Reply:
Maybe? File should contain the path to an existing directory on the device.

From [getExternalFilesDirs](https://developer.android.com/reference/android/content/Context#getExternalFilesDirs(java.lang.String))

- Shared storage may not always be available, since removable media can be ejected by the user. 
- The returned path may change over time if different shared storage media is inserted, so only relative paths should be persisted.

From [getExternalStoragePublicDirectory](https://developer.android.com/reference/android/os/Environment#getExternalStoragePublicDirectory(java.lang.String))

- Returns the File path for the directory. Note that this directory may not yet exist, so you must make sure it exists.

So, check if saved storage path has a parent directory that exists, if not then return the path of the primary shared storage.

For example in **File** `/storage/70C0-1C10/Download` check if **parentFile** `/storage/70C0-1C10` exists & isDirectory, if not return defaultStorage `/storage/emulated/0/Download` (usually)

DownloadManager will create the missing subdirectory `/Download` if necessary.

Reply:
I see. I think checking `/storage/70C0-1C10/Download` should already work, cause if `/storage/70C0-1C10` doesn't exist, `/storage/70C0-1C10/Download` won't exist as well. Should make the code a bit cleaner.

Reply:
The problem with doing it that way is that the `Download` directory [may not exist](https://developer.android.com/reference/android/os/Environment#:~:text=Note%20that%20this%20directory%20may%20not%20yet%20exist). And if the folder doesn't exist, the storage doesn't exist either.

- Without `parentFile` check (`/storage/70C0-1C10/Download`)
![sMCSQGl8o6](https://user-images.githubusercontent.com/25919231/106029433-b4d0b300-6092-11eb-8480-8347a99a9fe3.gif)

- With `parentFile` check (`/storage/70C0-1C10`)
![vlUW6ztt50](https://user-images.githubusercontent.com/25919231/106029553-d336ae80-6092-11eb-92f1-99c1342ad4f3.gif)

- Download Manager will create the missing subdirectory `/Download` if necessary.
![dYWRFUFT9g](https://user-images.githubusercontent.com/25919231/106029690-f5c8c780-6092-11eb-9066-e1ae97bd146a.gif)